x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialIntersects.java

```diff
@@ -0,0 +1,274 @@
+        return GEO.pointRelatesGeometry(leftValue, geometry);
+    }
+
+    @Evaluator(extraName = "GeoPointDocValuesAndSource", warnExceptions = { IllegalArgumentException.class })
+    static boolean processGeoPointDocValuesAndSource(long leftValue, BytesRef rightValue) {
+        Geometry geometry = SpatialCoordinateTypes.UNSPECIFIED.wkbToGeometry(rightValue);
+        return GEO.pointRelatesGeometry(leftValue, geometry);
+    }
+
+    @Evaluator(extraName = "CartesianSourceAndConstant", warnExceptions = { IllegalArgumentException.class, IOException.class })
+    static boolean processCartesianSourceAndConstant(BytesRef leftValue, @Fixed Component2D rightValue) throws IOException {
+        return CARTESIAN.geometryRelatesGeometry(leftValue, rightValue);
+    }
+
+    @Evaluator(extraName = "CartesianSourceAndString", warnExceptions = { IllegalArgumentException.class, IOException.class })
+    static boolean processCartesianSourceAndString(BytesRef leftValue, BytesRef rightValue) throws IOException {
+        BytesRef geometry = SpatialCoordinateTypes.UNSPECIFIED.wktToWkb(rightValue.utf8ToString());
+        return CARTESIAN.geometryRelatesGeometry(leftValue, geometry);
+    }
+
+    @Evaluator(extraName = "CartesianSourceAndSource", warnExceptions = { IllegalArgumentException.class, IOException.class })
+    static boolean processCartesianSourceAndSource(BytesRef leftValue, BytesRef rightValue) throws IOException {
+        return CARTESIAN.geometryRelatesGeometry(leftValue, rightValue);
+    }
+
+    @Evaluator(extraName = "CartesianPointDocValuesAndConstant", warnExceptions = { IllegalArgumentException.class })
+    static boolean processCartesianPointDocValuesAndConstant(long leftValue, @Fixed Component2D rightValue) {
+        return CARTESIAN.pointRelatesGeometry(leftValue, rightValue);
+    }
+
+    @Evaluator(extraName = "CartesianPointDocValuesAndString", warnExceptions = { IllegalArgumentException.class })
+    static boolean processCartesianPointDocValuesAndString(long leftValue, BytesRef rightValue) {
+        Geometry geometry = SpatialCoordinateTypes.UNSPECIFIED.wktToGeometry(rightValue.utf8ToString());
+        return CARTESIAN.pointRelatesGeometry(leftValue, geometry);
+    }
+
+    @Evaluator(extraName = "CartesianPointDocValuesAndSource")
+    static boolean processCartesianPointDocValuesAndSource(long leftValue, BytesRef rightValue) {
+        Geometry geometry = SpatialCoordinateTypes.UNSPECIFIED.wkbToGeometry(rightValue);
+        return CARTESIAN.pointRelatesGeometry(leftValue, geometry);
+    }
+
+    @Evaluator(extraName = "ConstantAndConstant", warnExceptions = { IllegalArgumentException.class, IOException.class })
+    static boolean processConstantAndConstant(@Fixed GeometryDocValueReader leftValue, @Fixed Component2D rightValue) throws IOException {
+        return CARTESIAN.geometryRelatesGeometry(leftValue, rightValue);
+    }
```

Comment:
Unless geo has some different meaning of the word "constant", I've never seen us have to do this before. The folding infrastructure should kick in if both sides are constant and we should use whatever evaluator is appropriate for the types and it'll be replaced by a constant during the optimization step. But it still runs with a regular evaluator.

Reply:
Which part are you commenting on? The use of `@Fixed`, which is something I've seen all over the place when functions can accept foldable types. Or the fact that the left is a `GeometryDocValueReader` and the right is `Component2D`, while the non-constant methods use `BytesRef` encoded WKB? For the second question, the process of figuring out if two geometries are related is to convert one to doc-values (embedded in a `GeometryDocValuesReader` and the other to a `Component2D`, to mimic what Lucene does. For constants we want that conversion done at planning time, hence the use of these types in this method, but for non-constants we will see `BytesRef` WKB formatted geometries coming out the blocks, and need to convert them deeper down the stack.

Reply:
OK. I see the `fold()` method, where I was foolishly just called super.fold, but now do the folding directly there, and can remove all these `ConstantAndConstant` methods and generated classes. Definitely an important cleanup.

Reply:
OK. That is pushed and working. Definitely much better to use `fold()` directly. Sorry for not spotting this myself!